rpcs3/ds3_pad_handler.cpp

```diff
@@ -0,0 +1,487 @@
+				thepad->m_port_status = CELL_PAD_STATUS_CONNECTED | CELL_PAD_STATUS_ASSIGN_CHANGES;
+				LOG_WARNING(HLE, "[DS3] Pad was connected");
+
+				connected++;
+			}
+
+			if (m_dev->large_motor != thepad->m_vibrateMotors[0].m_value || m_dev->small_motor != thepad->m_vibrateMotors[1].m_value)
+			{
+				m_dev->large_motor = thepad->m_vibrateMotors[0].m_value;
+				m_dev->small_motor = thepad->m_vibrateMotors[1].m_value;
+				send_output_report(m_dev);
+			}
+
+			break;
+		case DS3Status::Disconnected:
+			if (m_dev->status == DS3Status::Connected)
+			{
+				m_dev->status = DS3Status::Disconnected;
+				thepad->m_port_status = CELL_PAD_STATUS_DISCONNECTED | CELL_PAD_STATUS_ASSIGN_CHANGES;
+				libusb_close(m_dev->handle);
+				m_dev->handle = nullptr;
+				LOG_WARNING(HLE, "[DS3] Pad was disconnected");
+
+				connected--;
+			}
+			break;
+		}
+	}
+}
+
+void ds3_pad_handler::TestVibration(const std::string& padId, u32 largeMotor, u32 smallMotor)
+{
+	std::shared_ptr<ds3_device> device = get_device(padId);
+	if (device == nullptr || device->handle == nullptr)
+		return;
+
+	// Set the device's motor speeds to our requested values 0-255
+	device->large_motor = largeMotor;
+	device->small_motor = smallMotor;
+
+	int index = 0;
+	for (int i = 0; i < MAX_GAMEPADS; i++)
+	{
+		if (g_cfg_input.player[i]->handler == pad_handler::ds3)
+		{
+			if (g_cfg_input.player[i]->device.to_string() == padId)
+			{
+				m_pad_configs[index].load();
+				device->config = &m_pad_configs[index];
+				break;
+			}
+			index++;
+		}
+	}
+
+	// Start/Stop the engines :)
+	send_output_report(device);
+}
+
+void ds3_pad_handler::GetNextButtonPress(const std::string& padId, const std::function<void(u16, std::string, std::string, int[])>& callback, const std::function<void(std::string)>& fail_callback, bool get_blacklist, const std::vector<std::string>& buttons)
+{
+	if (get_blacklist)
+		blacklist.clear();
+
+	std::shared_ptr<ds3_device> device = get_device(padId);
+	if (device == nullptr || device->handle == nullptr)
+		return fail_callback(padId);
+
+	return;
+}
+
+void ds3_pad_handler::send_output_report(const std::shared_ptr<ds3_device>& ds3dev)
+{
+	u8 report_buf[] = {
+		0x00, 0xff, 0x00, 0xff, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0x00, 0x00, 0x00, 0x00, 0x00
+	};
+
+	report_buf[2] = ds3dev->large_motor;
+	report_buf[4] = ds3dev->small_motor;
+
+	libusb_control_transfer(ds3dev->handle, LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE, HID_SETREPORT, HIDREPORT_OUTPUT | 01, 0, report_buf, sizeof(report_buf), 0);
+}
+
+std::shared_ptr<ds3_pad_handler::ds3_device> ds3_pad_handler::get_device(const std::string& padId)
+{
+	if (!Init())
+		return nullptr;
+
+	size_t pos = padId.find(m_name_string);
+	if (pos == std::string::npos)
+		return nullptr;
+
+	int pad_number = std::stoi(padId.substr(pos + 9));
```

Comment:
Could've just done (padId[pos + 9] - '0') assuming no more than 9 controllers available at any one time, but this is fine too.